Added requirement for codesigning and timestamping certificate (#162)
notaryproject/notaryproject: signature-specification.md

### Certificate Requirements

> The codesigning and timestamping certificates MUST meet the following requirements:

Reviewer: ```suggestion
The codesigning and timestamping certificates MUST meet the following requirements. These requirements are validated both at signature generation time and signature verification time, and are applied to the certificate chain in the signature envelope. These validations are independent of certificate chain validation against a trust store.
```

1. **[Extended Key Usage:](https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.12)**  The `extendedKeyUsage` extension is OPTIONAL and can OPTIONALLY be marked as critical.
    - **For codesigning certificate:** If present, the value MUST contain `id-kp-codeSigning` and MUST NOT contain `anyExtendedKeyUsage`, `serverAuth`, `emailProtection` and `timeStamping`.
    - **For timestamping certificate:** If present, the value MUST contain `id-kp-timeStamping` and MUST NOT contain `anyExtendedKeyUsage`, `serverAuth`, `emailProtection` and `codeSigning`.
1. **Key Length** The certificate MUST abide by the following key length restrictions:
    - For RSA public key, the key length MUST be 2048 bits or higher.
    - For ECDSA public key, the key length MUST be 256 bits or higher.

#### Other requirements

1. The certificates in the signature MUST be ordered list of X.509 certificate or certificate chain i.e. the certificate containing the public key used to digitally sign the payload must be the first certificate, followed by the intermediate and root certificates in the correct order. This also means
    - The certificate MUST NOT chain to multiple parents/roots.

>     - The certificate chain MUST NOT contain a certificate that’s not part of the certificate chain.

Reviewer 2: Who is going to check for this condition?

Reviewer: ```suggestion
    - The certificate chain MUST NOT contain a certificate that’s is unrelated to the certificate chain.
```

> 1. A Valid certificate chain MUST contain a minimum of two certificates - a leaf and a root certificate.
> 1. The certificate in the certificate chain MUST NOT use SHA1WithRSA and ECDSAWithSHA1 signatures.

Reviewer: Nit
```suggestion
1. A valid certificate chain MUST contain a minimum of two certificates - a leaf and a root certificate.
1. Any certificate in the certificate chain MUST NOT use SHA1WithRSA and ECDSAWithSHA1 signatures.
```

> 1. Only Basic Constraints, Key Usage, and Extended Key Usage extensions of X.509 certificates are honored. For reset of the extensions, Notary MUST fail open i.e. they MUST NOT be evaluated and honored.

Reviewer: ```suggestion
1. Only Basic Constraints, Key Usage, and Extended Key Usage extensions of X.509 certificates are honored. For rest of the extensions, Notary MUST fail open i.e. they MUST NOT be evaluated or honored.
```

> 1. The certificate in the certificate chain MUST be valid at signing time. Notary MUST NOT enforce validity period nesting for certificates.

Reviewer 2: This is weirdly worded?

Author: Reworded it to `The certificates in the certificate chain MUST be valid at signing time. Notary MUST NOT enforce validity period nesting, i.e the validity period for a given certificate may not fall entirely within the validity period of that certificate's issuer certificate.`
